DVCodingStyle.md

```diff
@@ -594,30 +594,122 @@ cause race conditions if there are any parallel threads.
 
 ### Objections and Coordinating End of Test
 
-1.  At a high level, a phase should not end when either there is more stimulus
-    to send during that phase or the DUT has yet to respond to some stimulus that
-    has been sent.
-    To prevent a phase from ending in the former case, raise and lower objections
-    inside tests.
-    To prevent a phase from ending in the latter case, raise objections in the
-    scoreboard whenever expect queues are not empty and lower them when they are empty.
-2.  It is recommended to avoid using objections in drivers or monitors
-    for performance reasons they should be used inside these components
-    sparingly and only when needed.
-    (Note: this point is still under discussion, and is still TBD)
-3.  If a sequence needs to raise/drop objections, it should call
-    `raise_objection` in its `pre_start()` method and `drop_objection` in its
-    `post_start()` method. A good alternative to this is to call
-    `set_automatic_phase_objection(1)` in the sequence constructor.
-4.  Always perform the test `if (get_starting_phase() != null)` before calling
-    `raise_objection` or `drop_objection` within a sequence.
-5.  When starting a sequence that can raise and drop objections, if you want
-    the sequence to raise and drop objections, set the `starting_phase` member of
-    the sequence object the `<sequence_handle>.set_starting_phase(<uvm_phase>)`
-    before starting the sequence.
-6.  To avoid reimplementing objection functionality in multiple sequence classes,
-    create a parent sequence that implements this functionality for its derivatives.
-7.  When calling `raise_objection` or `drop_objection`, you must pass a string
+
+At a high level, a phase should not end when either there is more stimulus to
+send during that phase or the DUT has yet to respond to some stimulus that has
+been sent.
+
+To prevent a phase from ending in the former case, raise and lower objections
+inside the test class.
+
+To prevent a phase from ending in the latter case, it is required to raise and
+lower objections inside the `phase_ready_to_end()` function.
+It is also recommended to only include objections in the monitor component.
+When using this approach, these guidelines should be followed:
+
+1.  The monitor's `phase_ready_to_end()` function should implement a watchdog timer
+    that raises an objection if any traffic is seen on the bus and lowers it once
+    no traffic is seen within the timeout period; if traffic is seen, the
+    watchdog resets its timer.
+    Refer to the example below:
+
+    ```systemverilog
+    function void phase_ready_to_end(uvm_phase phase);
```

Comment:
In general, this is an enormous code snippet for a style guide. Would it be better to point at existing code in (say) OpenTitan as an example?

Reply:
I agree this is a big snippet. but I still like in the style guide as pointers to code need to be maintained.

Reply:
I tend to agree with @rasmus-madsen here, I agree that this is a large code snippet, but I also do prefer that it remains in the style guide rather than point to some other code files elsewhere in the OT repository; additionally due to the relative complexity of this watchdog mechanism, I feel having a "bare-bones" example here is more clear than pointing to an already-implemented monitor that might introduce some extra complexity into its implementation.

Reply:
Ok. Is there anything that can be cut out of this example? I mean, "It's really easy to do this, just write the following 50 lines of code" doesn't make our house style sound great :-)

Reply:
Hm, there's some code that gets and prints out any objects that still have raised objections that may not be the most necessary for this snippet (starting with the call to `phase.get_objection().get_objectors(objectors)`), and I did add numerous comments to explain what's going on.
@rasmus-madsen wdyt?

Reply:
I simplified the logic a bit, perhaps we can discuss it here https://github.com/lowRISC/opentitan/pull/2137, and then see what codes should be put here.

Reply:
That looks good to me. Do we suggest that code derives from `dv_base_...` classes? If so, maybe this style guide should say something like "derive from `dv_base_monitor`; poke `ok_to_end` as follows..." and we don't need to write out the implementation here.

Reply:
lowRISC/opentitan#2137 looks pretty concise in its implementation - I will update this example code here to match that structure as well.

Reply:
@rswarbrick w.r.t. derivation from the `dv_base` and `cip_lib` libraries, the style guide remains agnostic in this regard - it tries as much as possible not to make any assumptions about any DV library dependencies (apart from `dv_macros.svh` and `prim_assert.sv`).